pkg/operator/controller/status.go

```diff
@@ -95,123 +99,184 @@ func (r *reconciler) syncOperatorStatus() error {
 	} else {
 		progressingCondition.Status = configv1.ConditionTrue
 		progressingCondition.Reason = "Reconciling"
-		progressingCondition.Message = fmt.Sprintf("have %d daemonsets, want %d", numDaemonSets, numDNSes)
+		progressingCondition.Message = strings.Join(messages, "\n")
 	}
-	conditions = setStatusCondition(conditions, progressingCondition)
 
+	return setOperatorLastTransitionTime(progressingCondition, oldCondition)
+}
+
+// computeOperatorAvailableCondition computes the operator's current Available status state.
+func computeOperatorAvailableCondition(oldCondition *configv1.ClusterOperatorStatusCondition,
+	anyDNSesAvailable bool) configv1.ClusterOperatorStatusCondition {
 	availableCondition := &configv1.ClusterOperatorStatusCondition{
-		Type:   configv1.OperatorAvailable,
-		Status: configv1.ConditionUnknown,
-	}
-	daemonsetsAvailable := map[string]bool{}
-	for _, d := range daemonsets {
-		daemonsetsAvailable[d.Name] = d.Status.NumberAvailable > 0
-	}
-	unavailable := []string{}
-	for _, dns := range dnses {
-		name := DNSDaemonSetName(&dns).Name
-		if available, exists := daemonsetsAvailable[name]; !exists {
-			msg := fmt.Sprintf("no daemonset for dns %q", dns.Name)
-			unavailable = append(unavailable, msg)
-		} else if !available {
-			msg := fmt.Sprintf("daemonset %q is not available", name)
-			unavailable = append(unavailable, msg)
-		}
+		Type: configv1.OperatorAvailable,
 	}
-	if len(unavailable) == 0 {
+	if anyDNSesAvailable {
 		availableCondition.Status = configv1.ConditionTrue
+		availableCondition.Reason = "AsExpected"
+		availableCondition.Message = "At least 1 dns resource is available"
 	} else {
 		availableCondition.Status = configv1.ConditionFalse
-		availableCondition.Reason = "DaemonSetNotAvailable"
-		availableCondition.Message = strings.Join(unavailable, "\n")
+		availableCondition.Reason = "DNSUnavailable"
+		availableCondition.Message = fmt.Sprintf("No dns resources available")
 	}
-	conditions = setStatusCondition(conditions, availableCondition)
 
-	return conditions
+	return setOperatorLastTransitionTime(availableCondition, oldCondition)
 }
 
-// setStatusCondition returns the result of setting the specified condition in
-// the given slice of conditions.
-func setStatusCondition(oldConditions []configv1.ClusterOperatorStatusCondition, condition *configv1.ClusterOperatorStatusCondition) []configv1.ClusterOperatorStatusCondition {
-	condition.LastTransitionTime = metav1.Now()
+// checkAllDNSesAvailable checks if all DNS resources are available.
+func checkAllDNSesAvailable(dnses []operatorv1.DNS) bool {
+	for _, dns := range dnses {
+		available := false
+		for _, c := range dns.Status.Conditions {
+			if c.Type == operatorv1.OperatorStatusTypeAvailable && c.Status == operatorv1.ConditionTrue {
+				available = true
+				break
+			}
+		}
+		if !available {
+			return false
+		}
+	}
 
-	newConditions := []configv1.ClusterOperatorStatusCondition{}
+	return true
```

Comment:
We decided that a DaemonSet is available if 0/0 replicas are available (https://github.com/openshift/cluster-dns-operator/pull/88#discussion_r272673377, https://github.com/openshift/cluster-dns-operator/commit/5ac65e290667f001e1029fb0f69dc359875451c6#diff-c7469ee2a512d7eff942a36312cb4c50R20), and I see we are extending that reasoning by saying that all DNSes are available if 0/0 DNSes are available.

However, consider that the operator will automatically recreate the `default` DNS if it does not exist.  This consideration leads me to think that having 0/0 DNSes implies that `Progressing` is true, and possibly that `Available` is false.  We have discussed the analogous case of 0/0 ingress controllers in relation to the ingress operator: https://github.com/openshift/cluster-ingress-operator/pull/198#issuecomment-481058325.

Reply:
Given our invariant that `default` must always exist, agree that 0 `dns` resources means progressing=true.

IMO available should always be false if the actual replicas of `default` is 0 regardless of what the "desired". I don't understand how, from our operator's perspective, DNS could be considered available if no CoreDNS pods are running.

Reply:
>Given our invariant that default must always exist, agree that 0 dns resources means progressing=true.

Done, latest PR includes `Progressin=true` if `if dnsesAvailable == 0 || dnsesAvailable != dnsesTotal`.

Reply:
When there are no dns resources, I was expected these conditions:
Availabe = False --> No DNS is running and serving the requests
Progressing = False --> Operator is not actually doing any work or trying to transition from one state to another.
Degraded = True --> Operator expects default dns to be present but does not exist which is a failure.
@ironcladlou @Miciah What do you think?

Reply:
> Availabe = False --> No DNS is running and serving the requests

I agree.

> Progressing = False --> Operator is not actually doing any work or trying to transition from one state to another.

I disagree because the operator calls `ensureDefaultDNS` in a periodic goroutine, so the operator is doing work (or will be doing work in less than 1 minute) to create the default resource.

> Degraded = True --> Operator expects default dns to be present but does not exist which is a failure.

I agree.

Reply:
That makes sense. Okay, progressing=true when there are no dns resources.